ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,70 @@
+# 27. Availability Probe Framework
+
+* Date 2023-07-06
+
+## Status
+
+Accepted
+
+## Context
+
+As a product team, we want a composite system availability metric for both Production
+and Staging environments, so that we can make roadmap decisions with easy-to-understand
+feedback about the reliability of the system in mind.
+  
+This metric should tell us the overall up time of RHTAP with respect to services under
+the control of Red Hat.
+
+RHTAP is defined to be available at a given moment if all of its components are
+reporting to be available at that given moment.
+
+A component is defined to be available at a given moment if all of its availability
+probes are reporting to be available at that given moment.
+
+A convention is required for providing the availability of a probe.
+
+Once this is in place, those metrics can be aggregated in order to report the overall
+availability of RHTAP.
+
+## Decision
+
+Probes' availability will be provided as a Prometheus metric. The metric will contain
+labels to allow differentiating between the different probes. When exported out of the
+cluster of origin, additional labels will be attached to the metric to mark the cluster
+in which the metric was generated.
+
+### Delivery Method
+
+It is up for each team to define what it means for its component(s) to be available.
+Fundamentally, a component should be reported as available as long as it's capable of
+providing the service it aims to provide, and unavailable otherwise.
+
+Each team will define CronJobs that will test the availability of their components.
```

Comment:
Let's say that we have 30 components. So we will endup with 30 Cronjobs which will create job/pod every eg. minute?

When it's already decided (in decision section) that availablity probes are Prometheus metrics, and since prometheus is already configurated to aggregate metrics from each component why we don't have just one cronjob/service which would check all services and provide aggregated result?

Reply:
> and since prometheus is already configurated to aggregate metrics from each component

What are you referring to here?

The approach of having a single check that will examine RHTAP's availability end to end is something that was considered during [refinement](https://docs.google.com/document/d/1ar8PJX4MWmw7fpEgB4cxK7d8lXOpkojdTKrJPsok1A4/edit?usp=sharing), and eventually it was decided, at least for now, to go on something closer to blackbox monitoring. The cronjob approach is now suggested as a way to simplify the requirement for the individual teams so they don't have to start writing exporters.

I would say, that for a given service, if it's easier and granular enough to have a single job that will check the availability of multiple components within it, then I think that should be good enough.

> Let's say that we have 30 components. So we will endup with 30 Cronjobs which will create job/pod every eg. minute?

Yes. The frequency is something to be tweaked based on how much time it takes the job to complete and how frequent we want the status to be reported, given the resources the test consumes, but yes, that's the suggested idea.

Reply:
That was not the intention. The idea is for the cronjob to run some sort of checks against the component that will evaluate its availability (e.g. try to do something that the component knows to provide and report if it was able to do it or not).

Do you have in mind a case in which a component/service already reports metrics that can be translated into an availability probe (up/down indicator)?

Reply:
That confuses me, in decision section is that availability is provided by Prometheus metric. But now I can see that we should use CronJob which can receive availability even by different way.

Reply:
The idea of the framework is to let teams write cronjobs and have the o11y team translate them to metrics,  instead of having the teams writing test services that export metrics + service monitors to collect those metrics.

The team writes cronjob(s) that test the availability of their components (e.g. trying to trigger something within the component and verifying it was completed successfully).

The o11y team translates the results (success/failure) of all relevant cronjob runs and convert that to the availability metric.

So, the availability is provided by a Prometheus metric, and to deliver that the different teams need to come up with cronjobs to test the availability of their components, and the o11y will come up with means to transform those results in to the required metric.

Let me know if that makes sense, and also let me know if you can think of a way to make it clearer within the ADR itself.

Reply:
Won't these cron jobs be fairly expensive, e.g. in the JVM Build Service case the only really accurate way to tell if it is working as expected is to actually build something. I would image the other components are similar, in that the only way to actually get a reliable metric is to perform the actual action for real.

Reply:
@stuartwdouglas if that is the only way to tell the service is available then I think we should do it. It might be that the question in such case should be how frequent does it makes sense to run such check.

Reply:
I am definitely not against it. I just thought it was something that should be called out.

Reply:
Hopefully a single, occasional run of any given service would be insignificant next to the amount of times it is running for our users/customers.

Reply:
I think it is probably ok as is